### ISSUE TITLE
Numeric range requirements

### ISSUE THREAD

**Issue by kengruven** [kstenerud/concise-encoding#15]
There's one paragraph of the Spec which is providing me with more grief, as a user, than any other:

> "The Concise Encoding format itself places no bounds on the range of numeric types, but implementations (being bound by language, platform, and physical limitations) must decide which ranges to accept (as a library limitation and via schemas). Any numeric value that exceeds those ranges must be converted to positive or negative infinity. Infinity replacement must be done BEFORE other validation such as duplicate checking in containers."

I have so many questions!

- Infinity is defined only for floating point, not integers.  So a parser may choose to convert some integers to floats?

   Are there any other type conversions the parser may silently perform?  (Only one other that I've found so far: custom types to `@na` if unknown.)
- A large integer is very different from infinity.  If we really can't require bignums, why not use a stand-in like `@na:parser-overflow` for this?
- Is there no required range at all?  Can a valid parser choose to only accept numbers up to, say, ±100?
- Booleans and UUIDs are also classified as "numeric types" in the spec.  Is a parser allowed to accept only a subset of booleans/UUIDs, and overflow the rest to +Infinity, as well?
- Typed arrays contain only fixed-bit-width numbers.  When an out-of-bounds (to the parser) integer is found in an array, is the entire array promoted(/demoted) to a float array for this feature?

   Alternatively, is this subtly implying that parsers must always support numbers of at least 64 bits (or 128 bits?), so that such a contradiction could never occur?
- As overflow checks are defined to happen before distinctness checks, the number of elements in a map depends on what range the parser chooses to allow.  For example, if I send this data:

   ```
   c1 { favorite_primes = {
       34_100_539_960_937_404_223 = C
       17_102_107_120_689_194_899 = A
       76_298_768_165_666_329_973 = B }
   }
   ```
   then the program receiving it might say:
   ```
   "You've rated ${favorite_primes.keys.count} prime numbers!"
   ```
   and the user could see any number from 1 to 3.

   Do I need to use strings for numbers that I don't want to risk being corrupted?

That paragraph is especially strange to me because it stands in such contrast to the rest of the spec.  Overall, CE looks like a fairly maximally-featured format.  Implementors are required to support Unicode (including an up-to-date "confusables" database), datetimes (to nanosecond precision, including "proleptic Gregorian" and the IANA time zone database), bfloat16 (useful but still rare so I'll have to implement it by hand), LEB128, "compact-float" and "compact-time" (both completely custom), multiple collection types, UUIDs, and so on.  Yet bigints (which are standard in lots of languages, and have a million third-party implementations for the rest) are where we draw the line?

Clearly, being minimal and quick to implement (like JSON) is not a concern.  Nobody's writing CE in plain ANSI C, nor would they expect to be able to use only data types from libc if they did.  But in the case of integers (only), an implementor can simply choose to throw away data they don't want to handle.

(In case it isn't clear, I think that paragraph in the spec is kind of nuts.  I'd rather CE just make bigints mandatory, and require at least 64 bits for floats.  It seems like it would be simpler and more reliable for everyone.  Or maybe I'm interpreting it incorrectly?)

**Comment by kstenerud** [comment]
The idea behind that paragraph was to make it possible to do at least a partial recovery if the receiving implementation can't handle the data being received. It definitely needs a little more attention...

Basically, for almost every type supported by CE, there are languages that do not support either the type or the range of that type. In such a case, what does one do?

I put that numeric type paragraph in long before I added the NA type, so maybe now `@na:"cannot convert 1234567890123456789012345678901234567890 to uint64"` would be a better approach. But then what about languages that don't support such null-like constructs? What if you're reading into a struct and the receiving type is int32?

But then again, you have languages like Javascript, which only have ieee754 float types to emulate integer types, so anything over 54 bits can lose information (although it would be approximated). Should 0xffffffffffffffff be allowed to get its lower bits truncated in Javascript (which is what JSON parsers do)? If you're sending ID numbers as 64-bit ints that would be bad. But if you're sending measurements, maybe it's not so bad. And if it's a value like 5.49e100000000, should Javascript error out, put in an NA construct, or substitute infinity?

Anyway, that's where I'm coming from originally with that paragraph: Can we recover from unsupported type or range errors without changing the meaning or interpretation of the data? And if so, under what conditions could or should we even try? And if not, what should we do then? Almost every language will have some part of the spec that it can't fully support natively, and maybe won't even want to support for a particular use case. About the only language I can think of that is versatile enough to roll with pretty much all of it is Python, but it would do it by foisting this decision upon the user (who maybe would WANT to be the one to make the decision)...

The most likely outcome of CE adoption would be that most people would send only a subset of the spec in circumstances where maximum portability is required (lists, maps, ints, floats, bool, string, arrays, limited to 64-bit), and save the more esoteric types and ranges for situations where they control the implementations of both endpoints. In the Go reference implementation I'm already silently throwing out comments, metadata and NA types when the destination is a struct (in the Unmarshal method).

TBH I'm not sure where to strike the balance. Data truncation ranges from a just plain bad idea to maybe-better-than-nothing, depending on your circumstance. Dropping of unsupported data/ranges causes its own issues if the receiving end is doing things like counting entries in a container.

Probably in the end the least messy and confusing approach would be to substitute some kind of `@na:conversion-error` thing where the implementation can't support the data, and leave the user to deal with whatever trouble that causes (what to do if your boolean type receives an `@na:conversion-error`)... I've already been bitten by the Javascript JSON integers greater than 54-bits truncation thing, and the silent truncation caused a whole host of problems in production, so it's probably better to be explicit about data loss rather than implicit. IEEE754 is tricky enough already without that monster in the mix!

**Comment by kengruven** [comment]
Hmm, I see.  I'm much more OK with IEEE754 floats rounding to infinity.  Many programming languages do that already (Ruby, Python, JS, Swift, etc), so while I don't think it's *good* behavior, it's at least not too surprising.  (I consider IEEE754 a legacy format which is better suited to computation than transmission/storage.  I've got lots of issues with IEEE754 -- but that's another story...)

Integers are different.  I don't know any language offhand which rounds large integers to infinity.  Also, you run into lots of special cases (above) when trying to support automatic integer->infinity conversion.

BTW, JS now has bigints, and every modern browser supports this (covering nearly 90% of users already): https://caniuse.com/?search=bigint

> What if you're reading into a struct and the receiving type is int32?

So, there are many possible types of libraries one could write for implementing a data serialization format.  You can have a DOM-style one (load everything into a special data structure), a SAX-style one (emit events as data is read), or an ORM-style one (map everything into existing data structures, using provided rules).

This particular question is only an issue with the latter kind, and in that case, the programmer's rules (or the library default) would define what to do.  It looks to me essentially the same as the question of what to do if you try to load a UUID into a string field, or an NA into a non-nullable field, or if the user types in 99999 as their age: "it depends".  I don't really want CE to have to define every possible mapping of "what if you try to put data X in field of type Y".

The security issue is one I hadn't considered.  I would think the "Maximum coefficient size" would mostly handle that, though.  Maybe extend that to cover maximum total number of characters in a number or something, too.

**Comment by kstenerud** [comment]
OK, there's confusion in what I was trying to convey in the sentence "Any numeric value that exceeds those ranges must be converted to positive or negative infinity." :P It was intended as offering a solution to the problem of decoding a value that the receiver can't handle in its current form (because the value is out of the range that the decoder can handle or the structure can handle, or the decoder doesn't have the appropriate type, etc), and shouldn't have been a "must". There's no limit to what can be placed in any field. Whether it's an integer or a float is up to the sender, and its validity depends on the application (the schema would determine what is valid). But really, how the end application deals with this scenario isn't something that can be prescribed, because I have no way of knowing what their circumstances are, what data loss they can handle, etc.

So I'll just take that whole section out and put in a recommendations section at the end of the document for best practices when these situations do occur, or at least a description of the potential hazards of various solutions (rounding to infinity might make the data meaningless, truncating might change the meaning of values like identifiers, truncation or overflow could cause map key collisions, dropping of out-of-range data could impact code that counts number of entries, etc).

I've already got a section that touches on security issues, so this would just be in the same vein of recommendations. Things like map collisions are security issues anyway (JSON is particularly bad when it comes to security vulnerabilities, and CE would also have issues if people aren't made explicitly aware of what can go wrong).

**Comment by kengruven** [comment]
Ahh, I see.  Yes, I think I was interpreting it in a different way than you intended.  It sounded to me like "Users, be careful, because libraries may mess up your data", while it now sounds like it was meant to be more like "Library (and application) developers, be careful, because users are trusting you not to mess up their data".

(I think in the distant future this could eventually be solved by having separate guides for Users, and for Library Implementors, and for Application Developers.)

I'll give my OK to whatever recommendations you make here.  Thanks!

**Comment by kstenerud** [comment]
Actually it was mainly a placeholder for something intended to be a lot more involved than that... Here's a sneak peek (still a bit rough), but basically it was going to be a combination of data format security crash course, rules for library writers, and best practices for library users, since this subject transcends all of it:

Security and Limits
-------------------

Accepting data from an outside source is always a security risk. It's safest to always assume hostile intentions when ingesting data.

Although Concise Encoding supports a wide range of data types and values, any given implementation will have limitations of some sort on their abilities due to the platform, language, system, and performance profiles. This will inevitably lead to subtle differences in CE implementations and applications that an attacker might be able to take advantage of if you're not careful.


### Attack Vectors

There are many vectors that attackers can take advantage of when they control the data your system is receiving, the most common of which are induced data loss, field omission, key collisions, and exploitation of algorithmic complexity.

#### Induced Data Loss

Once the characteristics of a system are known, an attacker can anticipate under what circumstances it will suffer data loss based on the language and technology used, the host it's running on, and the code itself.

The most common sources of deserialization data loss are:

 * Cutting off the upper part of an integer on overflow
 * Rounding off the lower part of a binary float on overflow
 * Rounding to infinity or converting to NaN
 * Truncating long arrays/strings
 * Erasing/replacing invalid characters in a string-like object

Any point in your system that accepts data loss is a potential security hole, because different parts of your system will likely handle the same loss-inducing data in different ways, and those differences can be exploited by an attacker using specially crafted documents.

As an example, consider a fictional system where the access control subsystem leaves bad characters as-is or replaces them with u+fffd, and the storage subsystem truncates bad characters. If an attacker is able to send a "create user" or "change user" command with a group of `admin\U+D800` (which would pass access control validation because `admin\U+D800` != `admin`), he could set up an admin user because the storage subsystem truncates bad characters and stores the group `admin\U+D800` as `admin`. The next time that user is loaded, it will be in group `admin`.

Numbers can also suffer data loss depending on how the decoded values are stored internally. For example, attempting to load the value 0x123456789 into a 32-bit unsigned integer would in many languages silently overflow to a result of 0x23456789. Similarly, the value 0x87654321 (2271560481) stored in a 32-bit binary float field would be silently truncated to 2271560448, losing precision and changing the effective value because it only has 24 bits available for the siginificand.

#### Default Type Conversions

Default type conversions are a form of data loss, but are specially cited here because they tend to happen silently as a result of how the language they're implemented in works. For example, PHP silently converts string values to 0 when comparing to a number. C and C++ treat all nonzero values as "true", including pointers. Other languages convert the numeric value to a string and then compare. These differing behaviors can potentially be exploited by an attacker.

Default truncation is also a problem. For example, some systems will truncate out-of-range values to 0, meaning that if an attacker could somehow cause an out of range price (999999999999999999999999999999999999 or whatever), such a system would convert it to 0, giving it away for free!

#### Induced Omission

Omission occurs when a decoder decides to drop bad data rather than reject it, ignore it, or attempt to "fix" it. When this occurs, the count of objects in containers will change, and critical information might now be missing. This could become a security hole if the fields are preserved by the decoders in some parts of your system but omitted in others.

#### Key Collisions

Key collisions occur when two or more identical keys are present in the same encoded map. Systems could potentially deal with this situation in a number of different ways:

 * Take the initial key and reject all duplicate keys (choose the first instance)
 * Replace the value for each duplicate encountered (choose the last instance)
 * Reject the key and all associated values (choose nothing)
 * Reject the entire document (abort processing)

For example, given the map:

```
{
    purchase-ids = [1004 102062 94112]
    total = 91.44
    total = 0
}
```

As a seller, you'd want your billing system to choose the first instance of "total". As a buyer, you'd much prefer the second!

**Note**: Key collisions can also occur as a result of [data loss](#data-loss):

```
{
    purchase-ids = [1004 102062 94112]
    total = 91.44
    total\U+D800 = 0
}
```

In this case, should the system truncate bad Unicode characters after checking for duplicate keys, it would be vulnerable to exploitation.

#### Deserialization Complexity

Depending on the implementation, some operations can get expensive very quickly the larger the object is, exhibiting O(n²) or sometimes even O(n³) behavior. This is particularly true of "big int" type structures in many languages. Even attempting to deserialize values as small as 10^1000 into a BigInt could DOS such systems.

#### Other Vulnerabilities

Attack documents can threaten a system in other ways, too. For example:

 * Extremely large objects (for example a 1TB array).
 * Documents with too many objects.
 * Documents with too much container depth (attempting to overflow the decoder's stack).


### Mitigations: Concise Encoding Codecs

To mitigate these kinds of security issues, Concise Encoding codecs have the following additional requirements:

#### Data Error Events

When a data error occurs (validation of an object), a decoder must inform the application by replacing the data event with a specially crafted NA event that has the following map and fields as a reason:

```cte
@na:{
    reason = "description of why the value was rejected"
    value = "CTE encoding of the offending value, or @na:size if the value is beyond the size limits"
}
```

**Note**: If value is a string, and the operating environment has a tendency to automatically "fix" UTF-8 strings (by omission or replacement), encode it as a uint8 array instead.

If an NA event is received in place of a map key, it indicates that the entire key-value pair cannot be used.

#### User-Controllable Limits

The codec must allow the user to control various limits and ranges, with sane defaults to guard against DOS attacks:

 * Maximum object count overall
 * Maximum object count in a container
 * Maximum array length in bytes
 * Maximum total document size in bytes
 * Maximum depth from the top level (1) to the most deeply nested object (inclusive)
 * Maximum number of digits allowed in an integer value
 * Maximum number of significant digits allowed in a floating point number

It's impossible to prescribe what limits should be reasonable for all decoders, because different systems will have different constraints, and system capabilities increase as technologies improve.

As an illustration, for a general purpose decoder the following defaults should give a reasonable balance (in 2020):

| Metric                 | Limit     |
| ---------------------- | --------- |
| Document size in bytes | 10GB      |
| Overall object count   | 1 billion |
| Depth                  | 1000      |
| Coefficient size       | 100       |

#### Automatic Rejection

The codec must allow the user to control whether the codec automatically rejects the entire document when it encounters or produces an `@na` event while decoding, with a default of true.

#### Key Collision Behavior

The codec must allow the user to control whether a duplicate key is treated as a structural error (which halts procesing) or a data error (which replaces the data with an `@na` construct).

#### Validation

Values must be validated by the codec for:

 * [Global limits](#user-controllable-limits)
 * Content rules (based on type)
 * Schema rules (if any)

If an error is encountered while decoding:

| Error Type              | Procedure                         |
| ----------------------- | --------------------------------- |
| Structural errors       | Halt processing with a diagnostic |
| Structural limit errors | Halt processing with a diagnostic |
| Data validation errors  | Replace with @na event            |
| Data limit errors       | Replace with @na event            |
| Unsupported type        | Replace with @na event            |
| Duplicate key           | Halt or replace (user setting)    |

NA event construct to be replaced for value when data errors:


### Mitigations: Application Guidelines

 * Harmonize limits across your entire application surface (all subsystems).
 * Received values are all-or-nothing. If it can't be stored in its entirety without data loss, it should be rejected. Allowing data loss means opening your system to key collisions.
 * Guard against unintentional default conversions (example: string to 0 or true).
 * Never store @na or attempt to convert @na to a default value. @na can occur in map keys, opening your system to key collisions or other security risks.
 * If you're omitting rejected values, remember that rejecting one part of a key-value pair means rejecting the entire k-v pair.
 * Generate diagnostic messages when @na is encountered. Silent failure is dangerous.
 * If you wish to recover from invalid unicode data, always replace bad characters with u+fffd. Never truncate.
 * Use a schema as a guard against invalid data, bad structure, and unsupported types.
 * When in doubt, toss it out. Bad data is a security risk.
 * It's generally safer to reject the entire document than to reject only a part.

**Comment by kengruven** [comment]
That all sounds pretty good to me.

FWIW, I still don't think "Data Error Events" make sense for Typed Arrays, but then again I don't know any (recent) platforms where any fixed-size sequence of bits is invalid as an int/float, so that may never come up.

Personally, I think bignums aren't as rare or performance-intensive as this makes them sound.  I tried 12^3456 in several environments (Lisp, JS, Emacs, Ruby, Python, etc) on my ancient laptop and they all showed the result instantly.  But it can't hurt to warn people to keep numeric limits in mind.

I'm not sure what "Never store @na" means here.  Is that just a suggestion for when receiving a {reason/value} `@na` from a CE parser?  In general I think that `@na` corresponds to null/nil, in most languages/databases, and that's often a useful value to store and transmit.

"When in doubt, toss it out. Bad data is a security risk." sounds slightly ambiguous at first (one value, or the whole document?) -- until the next line ("safer to reject the entire document").  A document that's missing one piece is also "bad data", in my mind.  I might consider combining these two points somehow.

But I definitely agree with the overall principle that if a parser can't process the data exactly, it should either: (a) convert to `@na`-with-reason and/or warn the application in some other way; or else (b) fail and abort.  And it sounds like the choice between these two is intended to be up to the application developer, which is nice.

I'm no longer worried about my numbers being arbitrarily mangled by a CE parser!

**Comment by kstenerud** [comment]
Hmm strange, I hit this bigint problem in go a few years back where attempting to add 10000 digits slowed things down horribly. But maybe it was just a problem in their runtime library. Alright, I'll just put in some thing about profiling to test common choke points or something like that.

It's also looking like my quest to kill NULL has failed. NA can either be an in-band signaling mechanism for errors, or a NULL mechanism, but not both (it just gets too confusing). The original idea and intent was that wherever one might be tempted to put NULL, the same effect could be achieved by simply omitting the value from the document. Which is *almost* always true, except when you're doing an update of an existing object and want to break a link, for example. Omitting in that case could imply that you're simply not updating that field. So it looks like I'll probably have to reintroduce a NULL type to match the reality we live in, billion dollar mistake or not :/

So in that case, I'll probably rename 0x7e to `@nil`, and use `@na:@nil` `[94 7e 7e]` to indicate "not available for unknown reason". `@nil` will be for specifying "remove whatever was in this field on your side" (and also for "empty document"), and `@na:reason` will be for specifying "I couldn't provide this data for reasons", or for the decoder to say "I couldn't decode this value for reasons". Then the application developer can decide how to deal with it. Much of the confusion and chaos around NULL comes from ambiguity in its meaning anyway.

Or maybe I'm prescribing too much by forcing codec writers to inject `@na` on decode errors. There are language-specific ways to report errors after all, and they can probably be a lot more elegant...

Yeah, I'll reserve `@na` for over-the-wire error reporting since that's language agnostic by definition. Decoders can report data errors using their language's built-in mechanisms. In both cases, the app developer decides what to do about it. And so long as they use a consistent approach across their entire app, they shouldn't fall into security holes for it.

All updates will happen [here](https://github.com/kstenerud/concise-encoding/blob/master/ce-structure.md#security-and-limits).

**Comment by kengruven** [comment]
> when decoding bare numbers in CTE, you have to add a digit, multiply by 10, repeat for every digit encountered. As the number gets bigger, the slowdown is TREMENDOUS

Numeric libraries usually include [highly optimized parsers](https://github.com/python/cpython/blob/145bf269df3530176f6ebeab1324890ef7070bf8/Objects/longobject.c#L2183-L2267).  I experimented with parsing random base-10 integers in Python, and it's still a tiny fraction of a second at 100,000 digits.  (By 1,000,000 digits, it starts to bog down.)

It's true that if I iterate over the characters to multiply-and-add them myself, that's much slower.  So, don't write your own integer parser.  :-)  That approach may end up having performance issues with many small numbers, too.

Anyway, the default "coefficient size" in CE is suggested to be 100, which I think is probably big enough for everyone except mathematicians, and small enough to be several orders-of-magnitude away from being a DOS risk.

> Much of the confusion and chaos around NULL comes from ambiguity in its meaning anyway.

Right.  I think the problem with NULL is 50% that languages don't handle it well (which is gradually being solved), and 50% that it's semantically ambiguous.  Back in college, my database professor mentioned a paper where someone had investigated a bunch of real databases and catalogued 17 (or was it 27?) different meanings for "NULL".

I was actually rather impressed that CE had `@na`-with-reason, because it allows the possibility for me to clarify what my nulls mean!

I agree that using the same value for (intentionally missing) user data and in-band error signaling is a little weird.  (On the other hand, most languages still don't have great error handling support.  If everyone were using a language with a condition system like Lisp, we'd be in great shape!)  I'm playing with the idea of letting the user provide a delegate to control this, but I'm not sure exactly how it's going to work yet.

**Comment by kstenerud** [comment]
> Numeric libraries usually include highly optimized parsers. I experimented with parsing random base-10 integers in Python, and it's still a tiny fraction of a second at 100,000 digits. (By 1,000,000 digits, it starts to bog down.)

This is going to be up to the implementor, and will have to be on their judgment. I expect that this will take awhile to stabilize with each new implementation, as it's hard to discover what the real limitations are. Usually they're buried [deep within some library somewhere](https://github.com/cockroachdb/apd/blob/master/decimal.go#L73). Regardless, range errors when decoding is a [structural error](https://github.com/kstenerud/concise-encoding/blob/master/ce-structure.md#structural-errors), so there won't be any silent truncation.

> Right. I think the problem with NULL is 50% that languages don't handle it well (which is gradually being solved), and 50% that it's semantically ambiguous. Back in college, my database professor mentioned a paper where someone had investigated a bunch of real databases and catalogued 17 (or was it 27?) different meanings for "NULL".

Yes, I've been trying to avoid this, but it's surprisingly difficult. I've settled on a fairly [mundane meaning](https://github.com/kstenerud/concise-encoding/blob/master/ce-structure.md#nil) in the spec, which is hopefully enough to stave off the worst disasters!

[NA now always requires a reason, and implies an error condition](https://github.com/kstenerud/concise-encoding/blob/master/ce-structure.md#na). It's a bit ugly in that I had to set specific limits on where it can be placed, but this is to simplify the implementation by avoiding places where there should realistically be no NA object anyway...

As far as behavior goes, Nil represents an intentionally blank data field (generally for clearing an existing object of data in an update). It's expected that the client is ready to handle this during normal operations. NA represents an unintentionally blank data field (i.e. something went wrong, but maybe the receiving end can still use the overall data structure so that it's not a total loss).

The default behavior for all CE codecs will be to halt processing and reject the document whenever NA is encountered, with a user-defined config option to allow NA.
